c2rust-analyze/src/type_desc.rs

```diff
@@ -98,3 +99,169 @@ pub fn local_perms_to_desc<'tcx>(
+                steps.push(Step::Rc);
+                cur_ty = substs.type_at(0);
+            }
+            TyKind::Adt(adt_def, substs) if is_cell(tcx, adt_def) => {
+                steps.push(Step::Cell);
+                cur_ty = substs.type_at(0);
+            }
+            TyKind::Adt(adt_def, substs) if is_offset_ptr(tcx, adt_def) => {
+                steps.push(Step::OffsetPtr);
+                cur_ty = substs.type_at(0);
+            }
+
+            TyKind::Slice(inner_ty) => {
+                steps.push(Step::Slice);
+                cur_ty = inner_ty;
+            }
+            TyKind::Array(inner_ty, _) => {
+                steps.push(Step::Array);
+                cur_ty = inner_ty;
+            }
+
+            _ => panic!(
+                "failed to deconstruct {:?} as a pointer to {:?}: unexpected {:?}",
+                ty,
+                pointee_ty,
+                cur_ty.kind()
+            ),
+        }
+    }
+
+    let mut i = 0;
+    let next = |i: &mut usize| -> Option<&Step> {
+        let s = steps.get(*i);
+        *i += 1;
+        s
+    };
+
+    let own = match next(&mut i) {
+        Some(&Step::Ref(Mutability::Not)) => {
+            if let Some(&Step::Cell) = steps.get(i) {
+                i += 1;
+                Ownership::Cell
+            } else {
+                Ownership::Imm
+            }
+        }
+        Some(&Step::Ref(Mutability::Mut)) => Ownership::Mut,
+        Some(&Step::RawPtr(Mutability::Not)) => Ownership::Raw,
+        Some(&Step::RawPtr(Mutability::Mut)) => Ownership::RawMut,
+        Some(&Step::Box) => Ownership::Box,
+        Some(&Step::Rc) => Ownership::Rc,
+        _ => {
+            // Un-consume this step.
+            i -= 1;
+            panic!(
+                "failed to deconstruct {:?} as a pointer to {:?}: \
+                    steps {:?} don't start with a pointer",
+                ty,
+                pointee_ty,
+                &steps[i - 1..]
+            );
+        }
+    };
+
+    let qty = match next(&mut i) {
+        Some(&Step::Slice) => Quantity::Slice,
+        Some(&Step::OffsetPtr) => Quantity::OffsetPtr,
+        Some(&Step::Array) => Quantity::Array,
+        _ => {
+            // Un-consume this step.
+            i -= 1;
+            Quantity::Single
+        }
+    };
+
+    assert!(
+        i == steps.len(),
+        "failed to deconstruct {:?} as a pointer to {:?}: got extra steps {:?}",
+        ty,
+        pointee_ty,
+        &steps[i..],
+    );
+
+    (own, qty)
+}
+
+/// If `ty` is `Cell<T>`, returns `Some(T)`.  Otherwise, returns `None`.
+fn unpack_cell<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Option<Ty<'tcx>> {
+    match *ty.kind() {
+        TyKind::Adt(adt_def, substs) if is_cell(tcx, adt_def) => Some(substs.type_at(0)),
+        _ => None,
+    }
+}
+
+/// Returns `true` if `adt_def` is the type `std::cell::Cell`.
+fn is_cell<'tcx>(tcx: TyCtxt<'tcx>, adt_def: AdtDef<'tcx>) -> bool {
+    // TODO
+    false
```

Comment:
replace these with `todo!()`?

Reply:
Returning `false` effectively disables the `Cell` (or `Rc` / `OffsetPtr`) case for now.  We can implement the proper checks later when they eventually become necessary.  Having these cases disabled will already cause a panic - calling `unpack_pointer_type` with `ty = &Cell<i32>` and `pointee_ty = i32` will be unable to peel off the `Cell` (because the `Cell` case never fires), and will panic ("failed to deconstruct ...") because it can't reach the `i32` pointee type.  Using `todo!` here would move the panic somewhere else, with a less helpful message.